### PR TITLE
Master

### PR DESCRIPTION
**Pull request by TomDijkema**
- Add institution (publisher) functionality for GBIF
- Add CSV write functionality for publishers data
- New folders for CSV (sources and written)
- Renamed files: main.py -> main_GBIF.py, test.py -> csv_functions.py

- Tweaks to existing functions

### CODE REVIEW

**Comment by samleeflang** [comment]
For next time check if you can get the format of the branch to something like: "feature/{minimal description of functionality}"
For a bug fix we can then use bug/{}
I searched for the git branching name conventions which this comes from but was unable to find it.
However if we do this consistently we should be fine

**Comment by samleeflang** [comment]
Add a .gitignore file to your repo (top level).
This will ensure that files that you don't want to commit (like the .idea folder) are passed in the git commit.
https://git-scm.com/docs/gitignore (can also check a repo of mine)

File: csv_files/sources/microchanges.csv [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+,,,,,,,,,,,,,,,,,,,,,
+,,,,,,,,,,,,,,,,,,,,,
+,,,,,,,,,,,,,,,,,,,,,
+,,,,,,,,,,,,,,,,,,,,,
+,,,,,,,,,,,,,,,,,,,,,
+,,,,,,,,,,,,,,,,,,,,,
+,,,,,,,,,,,,,,,,,,,,,
+,,,,,,,,,,,,,,,,,,,,,
+,,,,,,,,,,,,,,,,,,,,,
+,,,"instn codes in gbif and ROR, enter gbif instn code or change it in ROR",,,,,,,,,,,,,,,,,,
```
Comment (samleeflang) [inline csv_files/sources/microchanges.csv:205]: I would cut the last part

**Comment by samleeflang** [comment]
I would probably remove the written folder as this should fill itself when to code has been run. It is not part of the code and can change each time you run it (so it isn't stable). Maybe have the directory there but have it empty.

**Comment by samleeflang** [comment]
Add a README.md file (which is in Markdown language https://www.markdownguide.org/) To describe to a user how they can run the application

**Comment by samleeflang** [comment]
Add a requirements file (https://realpython.com/lessons/using-requirement-files/) to the project so the version of your libraries are fixed.

File: main_GBIF.py [hunk 2]
```diff
@@ -152,3 +158,90 @@ def gather_issues_flags() -> dict:
...
+    for publisher in response['results']:
+        i = 0
+
+        # Basis of record
+        for boc in basis_of_record:
+            publishers[publisher['filter']['publishing_org']]['totals'][boc] = publisher['values'][i]
+            i += 1
+
+        # While looping for publisher, find issues and flags
+        query = {
```
Comment (samleeflang) [inline main_GBIF.py]: You could make part of the query into a default into which you append the remaining fields. Now each of the queries have fields repeated such as network_key, advanced, limit, offset, these could go to a generic variable

File: csv_functions.py [hunk 3]
```diff
@@ -221,7 +221,115 @@ def write_issues_and_flags_monthly(issues_and_flags: dict):
...
+
+
+def write_institution_to_csv():
+    """ Receives data from the institution function in main.py
+        Calls on related functions to write the data to csv
+        :return: Calls on: write_institution_to_csv_totals and write_institution_to_csv_issues_and_flags
+    """
+
+    # Get data
+    publishers = main.get_institutions_data()
```
Comment (samleeflang) [inline csv_functions.py]: Thit points to the old import should be main_GBIF

**Comment by samleeflang** [comment]
Check if you can get a clear main() method which start the script and runs through all the methods. I now find it a bit hard to see how it is all connected. I thought all logic was in the main_GBIF file and the csv_functions only write to the csv.
Then the flow would be:
main_GBIF -> main function -> first data retrieval -> data logic -> data write -> second data retrieval -> logic -> write, etc...
You could even split up these four parts, one file with the main, one file with the retrieval, one with the logic, one with the write.
Splitting the logic from the retrieval would also help with writing unit tests later.

Note that for not this splitting up is not really necessary, make sure there is a main() which calls all the functions and makes it into a nice executable program and only when there is time left do the refactoring

File: csv_functions.py [hunk 4]
```diff
@@ -221,7 +221,115 @@ def write_issues_and_flags_monthly(issues_and_flags: dict):
...
+            publisher['name'],
+            publisher['gbif_id'],
+            publisher['ror_id'],
+            publisher['totals']['datasets']
+        ]
+
+        for bor in basis_of_record:
+            values[publisher['gbif_id']].append(publisher['totals'][bor])
+
+    print(values)
```
Comment (samleeflang) [inline csv_functions.py]: When you want to print text (which is a good idea if we need to debug later) use a logger. We haven't gotten a fixed format yet but you could check out the image-meta-data where I have one. You can then also log at different log levels:
- debug (only enabled in test environments)
- info (normal information, no issues)
- warning (issue but not breaking anything, just something that shouldn't have happend)
- error (something breaks or fails with an particular issue)

**Comment by TomDijkema** [comment]
> For next time check if you can get the format of the branch to something like: "feature/{minimal description of functionality}" For a bug fix we can then use bug/{} I searched for the git branching name conventions which this comes from but was unable to find it. However if we do this consistently we should be fine

I think you mean something like the following article states: https://gist.github.com/digitaljhelms/4287848? So when I would develop a new functionality, I will need to create a local branch of master that is named after the feature or id in TeamWork. This way it is easier to see how changes were implemented. They also call out Bug and HotFix branches.

**Comment by TomDijkema** [comment]
> Add a .gitignore file to your repo (top level). This will ensure that files that you don't want to commit (like the .idea folder) are passed in the git commit. https://git-scm.com/docs/gitignore (can also check a repo of mine)

I have added a .gitignore file to the following push for .idea and pycache, not sure if this will be part of this pull request or a new one (do not exactly know how this flow functions).

**Comment by TomDijkema** [comment]
> I would probably remove the written folder as this should fill itself when to code has been run. It is not part of the code and can change each time you run it (so it isn't stable). Maybe have the directory there but have it empty.

Do you mean the contents of the folder? At the moment the csv files are being overwritten each time. I recalled hearing that we could store the csv by year and month, would that be an executable idea?

**Comment by TomDijkema** [comment]
> Add a README.md file (which is in Markdown language https://www.markdownguide.org/) To describe to a user how they can run the application

Will be included in the next commit / push

**Comment by TomDijkema** [comment]
> Add a requirements file (https://realpython.com/lessons/using-requirement-files/) to the project so the version of your libraries are fixed.

Added requirements file for the next commit/push

**Comment by TomDijkema** [comment]
> Check if you can get a clear main() method which start the script and runs through all the methods. I now find it a bit hard to see how it is all connected. I thought all logic was in the main_GBIF file and the csv_functions only write to the csv. Then the flow would be: main_GBIF -> main function -> first data retrieval -> data logic -> data write -> second data retrieval -> logic -> write, etc... You could even split up these four parts, one file with the main, one file with the retrieval, one with the logic, one with the write. Splitting the logic from the retrieval would also help with writing unit tests later.
> 
> Note that for not this splitting up is not really necessary, make sure there is a main() which calls all the functions and makes it into a nice executable program and only when there is time left do the refactoring

I have simplified the main file so every function can be called from a core function called 'main'. Code and file logic has been adjusted to this approach. A future installment could also separate the data retrieval and logic.

File: .gitignore [hunk 5]
```diff
@@ -0,0 +1,3 @@
+.idea
```
Comment (samleeflang) [inline .gitignore:1]: Deze file kan weg, die op root niveau is voldoende

**Review by samleeflang — APPROVED** [review]

**Comment by samleeflang** [comment]
Starting to look good! I think we can tighten a couple of spot but let's do that in the next round of review. Let's let this pass so we can move on with the story and the task can move to the completed lane before the sprint planning

File: csv_files/sources/microchanges.csv (hunk 1 above)
Comment (TomDijkema) [reply csv_files/sources/microchanges.csv:205]: Done

File: main_GBIF.py (hunk 2 above)
Comment (TomDijkema) [reply main_GBIF.py]: Done

File: csv_functions.py (hunk 3 above)
Comment (TomDijkema) [reply csv_functions.py]: Done

**Pull request merged**